### PR TITLE
FIX: macosx flush_events should process all events

### LINKED ISSUE #23327: [Bug]: Event loop issue with Macosx backend
### Bug summary

The Macosx backend won't show a window until all the processing is completed, even when calling `plt.show()` beforehand. For a continuously updating figure, this means that it will just hang forever without displaying anything. I can do this fine in Tkagg, but I'd like to be able to display retina-quality images.

### Code for reproduction

```python
import matplotlib, random
import matplotlib.pyplot as plt

matplotlib.use('macosx') #Works fine with tkagg
data = []

#Initial Draw
fig, axes = plt.subplots(1)
line, = axes.plot(data, label='Some data', color='#CC0000')
fig.canvas.draw_idle()
plt.show(block=False)

while True:
	data += [random.randint(0,50) for i in range(20)]

	tseries = range(0, len(data))
	line.set_ydata(data)
	line.set_xdata(tseries)
	axes.relim()
	axes.autoscale_view(tight=False)

	if fig.stale: fig.canvas.draw()
	fig.canvas.flush_events() #Listen for user input
```

### Actual outcome

Python hangs forever, waiting for the `while` loop to complete.

### Expected outcome

As with Tkagg, it should launch the window first and continuously update the plot.

### Additional information

_No response_

### Operating system

Mac OS 10.15

### Matplotlib Version

3.5.2

### Matplotlib Backend

MACOSX

### Python version

3.10.5

### Jupyter version

_No response_

### Installation

pip

### PR DESCRIPTION
**Pull request by greglucas**
## PR Summary

Rather than just bursting the runloop for a short amount of time, we can run it until all pending events have been processed. In this example on main, the buttons are all unable to be pushed and close/minimize are grayed out. The buttons should all work with this update.

```python
import matplotlib, random
import matplotlib.pyplot as plt

matplotlib.use('macosx') #Works fine with tkagg
data = []

#Initial Draw
fig, axes = plt.subplots(1)
line, = axes.plot(data, label='Some data', color='#CC0000')
fig.canvas.draw_idle()
plt.show(block=False)

while True:
	data += [random.randint(0,50) for i in range(20)]

	tseries = range(0, len(data))
	line.set_ydata(data)
	line.set_xdata(tseries)
	axes.relim()
	axes.autoscale_view(tight=False)

	if fig.stale: fig.canvas.draw()
	fig.canvas.flush_events() #Listen for user input
```

closes #23327

## PR Checklist

**Tests and Styling**
- [N/A] Has pytest style unit tests (and `pytest` passes).
- [N/A] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Review by QuLogic — APPROVED** [review]
This seems consistent with what Tk does, https://github.com/tcltk/tk/blob/main/generic/tkCmds.c#L1201

**Comment by tacaswell** [comment]
One way I can see this going wrong is if in the process of processing the event it adds an event to the queue.  Will this cut of at "only events that were in the queue when we started" or will it get stuck?

https://developer.apple.com/documentation/foundation/nsdefaultrunloopmode should also win an award for likely correct but very unhelpful documentation.

File: src/_macosx.m [hunk 1]
```diff
@@ -394,9 +394,20 @@ static CGFloat _get_device_scale(CGContextRef cr)
...
-    // We need to allow the runloop to run very briefly
-    // to allow the view to be displayed when used in a fast updating animation
-    [[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.0]];
+    // We run the app, matching any events that are waiting in the queue
+    // to process, breaking out of the loop when no events remain and
+    // displaying the canvas if needed.
+    NSEvent *event;
+    while (true) {
+        event = [NSApp nextEventMatchingMask: NSEventMaskAny
+                                   untilDate: [NSDate distantPast]
```
Comment (tacaswell) [inline src/_macosx.m:403]: I do not fully understand https://developer.apple.com/documentation/appkit/nsapplication/1428485-nexteventmatchingmask . It is not clear if "expiration" here means "do not wait longer than this" or "do not return any events newer than this".  If it is the second, I think grabbing the current time outside of the `while` loop and passing it here will prevent the perpetual-motion machine I am worried about.

**Comment by greglucas** [comment]
I pushed up that idea in the most recent commit. It seems to still work. The only thing is that we are popping the event off the queue and resending them, so I assume that "resend" event may or may-not get processed and if it doesn't wouldn't meet the criteria the second go-around. So, is the purpose here to "flush" the queue and drain it entirely, or just to try and send all events... Another option would be to create a new queue entirely and keep track ourselves, but I think some of these can be future bug fixes too.

Here is a link to some QT event processing: https://code.woboq.org/qt5/qtbase/src/plugins/platforms/cocoa/qcocoaeventdispatcher.mm.html#462

File: src/_macosx.m (hunk 1 above)
Comment (greglucas) [reply src/_macosx.m]: See recent commit. The `distantFuture` discussion I think indicates the second of those. https://developer.apple.com/documentation/foundation/nsdate/1415385-distantfuture

**Comment by tacaswell** [comment]
I would trust the Qt code over my rambling.   There is also a question of concurency here that I am not sure of.  Does `[NSApp sendEvent:event];` (https://developer.apple.com/documentation/appkit/nsapplication/1428359-sendevent) do that work _now_ or schedule something on other objects that will happen the next time they get a chance to run?  If they try at add events to the queue we are pulling things off of does that happen _now_ or when we return from our function?

For the Qt code `processEvents` (https://doc.qt.io/qt-6/qcoreapplication.html#processEvents) should only process thing enqueue before the call so I think that means your initial commit is correct (and I do not understand this code).

File: src/_macosx.m (hunk 1 above)
Comment (tacaswell) [reply src/_macosx.m]: > You can use the object returned by distantFuture as the date argument to wait indefinitely for the event to occur.

I read this as pointing to the first (it is a time out not a filter)?

**Comment by greglucas** [comment]
I am out of my league here, I'm not truly a mac programmer ;)
I was basing some of this off of other event handlers within the backend as well.
https://github.com/matplotlib/matplotlib/blob/b54b3355ccb1eed0d24467de62d1e886948f38aa/src/_macosx.m#L149-L160

I still think the first commit is potentially better than what we had before because the previous version was non-responsive and not really doing the right thing. This one may have the potential to get caught in the never-ending loop if an event fires 10 new events and just keeps building the event queue up, but that is likely a low-likelihood case.

Regardless, I think all of this can be considered bug-fixes and brought in after the rc even.

**Comment by QuLogic** [comment]
Neither Tk nor GTK (and presumably wx along with it) special case that behaviour, so I guess we are inconsistent either way. IIRC, there was some discussion about this for the GTK4 port as well.

**Review by tacaswell — APPROVED** [review]

**Pull request merged**